vars/deployPlannerSnapshot.groovy

```diff
@@ -0,0 +1,67 @@
+#!/usr/bin/groovy
+
+import java.util.regex.Pattern
+
+def call(body) {
+    // evaluate the body block, and collect configuration into the object
+    def config = [:]
+    body.resolveStrategy = Closure.DELEGATE_FIRST
+    body.delegate = config
+    body()
+
+    def yaml
+    def openShiftProject = config.openShiftProject
+    def openShiftTemplate = config.openShiftTemplate
+    def originalImageName = config.originalImageName
+    def newImageName = config.newImageName
+    def deploymentName = config.githubRepo
+    def providerLabel = config.providerLabel ?: 'fabric8'
```

Comment:
confuses me as well, and follows what is done in https://github.com/fabric8io/fabric8-pipeline-library/blob/master/vars/deployOpenShiftSnapshot.groovy#L11 Is there some magic involved?

Reply:
as mentioned above in comment - https://github.com/fabric8io/fabric8-pipeline-library/pull/378/files#diff-e29e1891af9b97dd7f7137d912b468acR9 this line is responsible

Reply:
@jaseemabid wonder why you thought all local variables would be `null`? wouldn't `body()` initialise these?

Reply:
@pranavgore09 @jaseemabid hopefully this may make sense to you ... 

```groovy
class Body {

  def closure =  {
    println delegate
    println "closure: foo: ${foo}"
    foo = 'barrrr'
    bar = 'baaazzz'
  }
}


def config = [ foo: "bar"]
def obj = new Body()
def body = obj.closure

body.delegate = config

println "main: foo: ${config.foo}"
println "main: bar: ${config.bar}"

body()

println "main: foo: ${config.foo}"
println "main: bar: ${config.bar}"


```

Reply:
just in case you are wondering where I got this idea from  :) - https://stackoverflow.com/a/23450944